File: Trees/0110. Balanced Binary Tree/0110.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int height = 0;
        return DFS(root, height);
    }

private:
    bool DFS(TreeNode* root, int& height) {
        if (root == NULL) {
            height = -1;
            return true;
        }

        int left = 0;
        int right = 0;

        if (!DFS(root->left, left) || !DFS(root->right, right)) {
            return false;
        }
        if (abs(left-right) > 1) {
            return false;
        } 

        height = 1 + max(left, right);
        return true;
    }
};

File: Trees/0110. Balanced Binary Tree/0110.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        # must keep track of a height and a bool to check if node is balanced -> children must be balanced as well

        def dfs(root):
            if not root:
                return [0, True]
            
            left = dfs(root.left)
            right = dfs(root.right)

            